Watchman.DomainModel/Configuration/Services/ConfigurationService.cs

```diff
@@ -0,0 +1,115 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Autofac;
+using Watchman.Integrations.MongoDB;
+
+namespace Watchman.DomainModel.Configuration.Services
+{
+    public class ConfigurationService : IConfigurationService
+    {
+        private const ulong DEFAULT_SERVER_ID = 0;
+        private readonly IComponentContext _componentContext;
+        private readonly ISessionFactory _sessionFactory;
+        private readonly ConfigurationMapperService _configurationMapperService;
+        private readonly ConfigurationItemsSearcherService _configurationTypesSearcher;
+        private static Dictionary<Guid, int> _configurationVersions;
+        private static Dictionary<Type, Dictionary<ulong, IMappedConfiguration>> _cachedConfigurationItem;
+
+        public ConfigurationService(IComponentContext componentContext, ISessionFactory sessionFactory, ConfigurationMapperService configurationMapperService, ConfigurationItemsSearcherService configurationTypesSearcher)
+        {
+            this._componentContext = componentContext;
+            this._sessionFactory = sessionFactory;
+            this._configurationMapperService = configurationMapperService;
+            this._configurationTypesSearcher = configurationTypesSearcher;
+            this.Refresh();
+        }
+
+        public T GetConfigurationItem<T>(ulong serverId) where T : IMappedConfiguration
+        {
+            var configurations = _cachedConfigurationItem[typeof(T)];
+            var serverConfiguration = configurations.GetValueOrDefault(serverId) ?? configurations[DEFAULT_SERVER_ID];
+            return (T)serverConfiguration;
+        }
+
+        public IEnumerable<IMappedConfiguration> GetConfigurationItems(ulong serverId)
+        {
+            return _cachedConfigurationItem.Select(x => x.Value.GetValueOrDefault(serverId) ?? x.Value[DEFAULT_SERVER_ID]);
+        }
+
+        public async Task SaveNewConfiguration(IMappedConfiguration changedConfiguration) // todo: not saving new configuration not tested
+        {
+            using var session = this._sessionFactory.Create();
+            var existingConfiguration = session.Get<ConfigurationItem>()
+                .FirstOrDefault(x => x.ServerId == changedConfiguration.ServerId && x.Name == changedConfiguration.Name);
+            var baseFormatConfigurationItem = this._configurationMapperService.MapIntoBaseFormat(changedConfiguration);
+            if (existingConfiguration == null)
+            {
+                await session.AddAsync(baseFormatConfigurationItem);
+            }
+            else
+            {
+                existingConfiguration.SetValue(baseFormatConfigurationItem.Value);
+                await session.AddOrUpdateAsync(existingConfiguration);
```

Comment:
fajnie byłoby przeprojektować system z aktualizowaniem danych w bazie
w taki sposób, żeby aktualizowało te obiekty, które mają wyższe Version
ale bez pobierania obiektów w bazie 

może da sie zrobić to na etapie generowania zapytania do mongo, konfiguracji połączenia z bazą albo czegoś innego 

żeby ograniczyć ilość wysyłanych zapytań do mongo - które mongo musi przetwarzać, sprawdzać itd
a jak wiadomo, mamy darmową baze która jest wolna

w przypadku aktualizowania statystyk w statsmanie w handlerze dodałem sztywnego ifa bo akurat tam sie dało to zrobić i różnica między ignorowaniem obiektów w których nie ma co aktualizować i próbą aktualizowania wszystkeigo była ogromna 

może dobry pomysł to będzie properties/field (np prywatny field + metoda) 
który będzie trzymać informacje czy od czasu stworzenia/pobrania go z bazy został zmieniony 

i na etapie Update w Session będziemy sprawdzać tą wartość 

===========

podczas tworzenia obiektu lub pobierania go z bazy będzie zawsze ustawiony na false
a jakiekolwiek użycie metody Update() zmienia go na true